libs/core/synchronization/include/hpx/synchronization/latch.hpp

```diff
@@ -285,9 +284,40 @@ namespace hpx { namespace lcos { namespace local {
             HPX_ASSERT(old_count == 0);
             HPX_UNUSED(old_count);
 
-            std::unique_lock<mutex_type> l(mtx_.data_);
+            std::scoped_lock l(mtx_.data_);
             HPX_ASSERT(notified_);
             notified_ = false;
         }
+
+        /// Effects: Equivalent to:
+        ///             if (is_ready())
+        ///                 reset(count);
+        ///             count_up(n);
+        void count_up_or_reset(std::ptrdiff_t n, std::ptrdiff_t count)
```

Comment:
I suppose this is slightly better than the "manual" version in the comment above, because it avoids potentially taking the lock twice?

My only concern is the name, since it's `count_up_and_reset` when the `latch` is ready, and otherwise it's `count_up`. However, I don't have a suggestion for a better name so I'm only mentioning this in case you or someone else does.

Reply:
> I suppose this is slightly better than the "manual" version in the comment above, because it avoids potentially taking the lock twice?

Yes, this also avoids associated race conditions.

> My only concern is the name, since it's count_up_and_reset when the latch is ready, and otherwise it's count_up. However, I don't have a suggestion for a better name so I'm only mentioning this in case you or someone else does.

It's the best name I was able to come up with ;-)

Reply:
I like the suggestion you mentioned yesterday better than the current option, i.e. something like `reset_if_needed_and_count_up`.